design/authenticators_api.md

```diff
@@ -0,0 +1,178 @@
+# Introduction
+
+This doc challenges the current `/authenticators` endpoint in the Conjur API, and
+suggests a new approach.
+
+Full feature doc for the Authenticators Health-Check API can be found [here](authenticators_health_api.md)
+
+### Personas
+
+|     **Persona**    |                       **Description**                             |
+|:---------------:|:-----------------------------------------------------------------:|
+| Conjur Operator | A Conjur user with enhanced privileges (loads policies to root, configures authenticators, etc.) |
+| Conjur User     | A developer/app that needs to login to Conjur to retrieve secrets |
+
+## Current Mechanism
+
+Before we go over the suggestion for the new API, let's understand the flaws of
+the current mechanism. Currently, anyone can run an `/authenticators` request 
+and will get the following response:
+```
+{
+    "installed": [
+        "authn",
+        "authn-1",
+        "authn-2",
+        "authn-3"
+    ],
+    "configured": [
+        "authn",
+        "authn-1"
+    ],
+    "enabled": [
+        "authn",
+        "authn-1",
+        "authn-2"
+    ]
+}
+```
+
+Let's go over the sections in the response to understand them:
+
+1. installed: these authenticators are implemented in conjur code 
+(e.g they have an `Authn-somehting::Authenticator` class with a `valid?` method).
+
+2. configured: these authenticators have a webservice called 
+`conjur/authn-something/webservice`.
+
+3. enabled: these authenticators are whitelisted in the CONJUR_AUTHENTICATORS variable.
+
+Section 1 has important data but it shouldn't be here, but rather in the docs. It has
+nothing to do with the given Conjur instance and just describes the authenticators that
+are implemented in Conjur.
```

Comment:
I would disagree here as keeping docs and codebase in lock-step would be incredibly difficult. Ideally you could use this data to somehow enable new authenticators (even though we currently have no method for doing this).

With that said, I think we should have different permissions to see each part of the data:
- `1` - admin account only
- `2` - high-privilege account (operator that may or may not be the main admin account)
- `3` - any user (including unauthenticated ones)

Reply:
i mentioned this in the doc itself but having 2 & 3 in the current way is very confusing. I'm a conjur dev and it took me a while to understand what is what so a user will be confused for sure. in my suggestion we will have just one piece of data - is the authenticator available for authentication. _how_ we define what is available (configured + enabled in the env) is too much data imo.

Reply:
regarding 1 - you have the version in the `info` endpoint so you can go to the documentation of that version and check which authenticators are implemented in that version.

Reply:
I don't think we have versioned documentation right now, so I'm not sure in practice someone would be able to discover implemented authenticators this way.

Reply:
> having 2 & 3 in the current way is very confusing. 

This might be something that we should look into by separating administrative endpoints with a specific prefix (`<host>/api/foo/bar` vs `<host>/api/admin/foo/bar`)

> regarding 1 - you have the version in the info endpoint so you can go to the documentation of that version and check which authenticators are implemented in that version.

If you think about how much work it is to get the relevant info they need to get things working to the developer vs a simple rest endpoint, I don't think it's a good developer experience. I always get reminded of how terrible it is to look for OpenShift functionality and after failing to get something running for a while, you notice that the google search linked you to an ancient version of APIs.

Reply:
ok, here's another approach. **As a user**, all that matters to me is which authentication are available for authentication. i have nothing to do with configuration, so when i hit the `/authenticators` endpoint I'll get
```
{
   “authenticators”:
   [
      "authn",
      "authn-1"
   ]
}
``` 
even though `authn-2` is implemented in conjur in the current version but not configured.

**As an operator** i do care about which authenticators i can configure so when i hit the `/authenticators/status` endpoint i'll get all the authenticators that are implemented in the current version of conjur.

having all the authenticators in the general `/info` endpoint is still too much information imo (not security-wise but ux) and can mislead a user. as a user i'm not aware of what makes an authenticator available for configuration so if i see `authn-2` in the response i can easily think that i can authenticate with it.

@izgeri @sgnn7 wdyt?

Reply:
@izgeri @sgnn7 would love to get your comments here :)

Reply:
@orenbm I like this approach 👍